highfidelity/hifi
# Avatar bandwidth budget improvements

This is the beginning of sharing the avatar priority sort used in simulation logic with the avatar mixer for bandwidth budgeting...

This PR includes:
- moving the actual `sortAvatars` method to be a static of `AvatarData` . I'm open to suggestions for a better place for this... this was someplace that is common to the mixer and the client. It may not actually need access to the AvatarData class... so maybe putting it in it's own namespace would be more OO friendly (data hiding from `AvatarData` members)

- adds a message between client and avatar-mixer when the client uses the priority tweak nobs (this is temporary only for debugging/fine-tuning)...

- uses the avatar sort order to handle bandwidth budget

- stops sending avatar data in the frame once bandwidth budget is reached, this results in never going over budget and as a result the "distance based random drops" don't kick in.

**Test Plan:**
- Smoke test with this sandbox/domain and a couple of different agents... you must test multi-user support.
- Make sure ignore still works
- Make sure when you turn away from avatars that your inbound avatar mixer bandwidth goes down
- Make sure that when you turn away from avatars they don't "disappear from your avatar list"
- Load test with more than 20 avatars... make sure that animation still seems reasonable.
    - when you get above ~15 avatars, the 5MB bandwidth budget is insufficient to send all joints at 45hz, so the system has to start prioritizing
    - the current master code uses a random distance based technique to drop avatars...
    - this build will prioritize avatars by apparent size, closeness to the center of your view, and time since last update
    - the overall expected result is that you should see everyone moving (they shouldn't disappear and/or freeze for long periods of time)
    - generally things closer and more in the center of your view should have a higher update rate than things further and off to the side of your view
